backend/api/src/realms.rs

```diff
@@ -57,35 +61,43 @@ impl Tree {
                 "select id, name, parent from realms",
                 std::iter::empty(),
             ).await?;
+
         Ok(row_stream.map_ok(|row| Realm {
-            id: row.get(0),
+            id: id_from_sql(row.get(0)),
             name: row.get(1),
-            parent_id: row.get(2),
+            parent_id: id_from_sql(row.get(2)),
         }))
     }
 
     pub(super) async fn from_db(db: &Pool) -> anyhow::Result<Self> {
         // We store the nodes of the realm tree in a hash map
         // accessible by the database ID
         let realms = Self::raw_from_db(db).await?
-            .map_ok(|realm| (realm.id, realm))
+            .map_ok(|realm| (realm.id.key(), realm))
             .try_collect::<HashMap<_, _>>().await?;
 
         // With this, and the `parent` member of the `Realm`,
         // we already have quick access to the data of a realm's parent.
         // To also get to the children quickly we maintain another map.
-        let mut children = HashMap::<_, Vec<_>>::new();
+        let mut children = <HashMap<_, Vec<_>>>::new();
 
         for Realm { id, parent_id, .. } in realms.values() {
             if id != parent_id {
-                children.entry(*parent_id).or_default().push(*id);
+                children.entry(parent_id.key()).or_default().push(id.key());
             }
         }
 
         Ok(Tree { realms, children })
     }
 
     pub(super) fn get_node(&self, id: &Id) -> Option<&Realm> {
-        self.realms.get(id)
+        if id.kind() != KIND_PREFIX {
+            return None;
+        }
```

Comment:
I almost forgot this check. I kinda worry that the current API of `Id` makes this kind of bug easy. I wonder if we can improve anything about it. Maybe have `Id::key_for_kind(expected_kind) -> Option<u64>`?

Reply:
I changed `key` to `key_for` now. So far I think that's a fine solution.